File: past-competitions/2018/src/brett-cpp/Prob16.c
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int main(int argc, char *argv[])
{
	// open file for reading
	FILE *inputFile = fopen("Prob16.in.txt", "r");
	int testCases;

	// read number of test cases
	fscanf(inputFile, "%d\n", &testCases);
	
	int centerX, centerY, points, outerRadius, innerRadius;
	double currentAngle, angleSpacing, x, y;

	// execute test cases
	while(testCases > 0){
		// read inputs
		fscanf(inputFile, "%d %d %d %d %d\n", &centerX, &centerY, &points, &outerRadius, &innerRadius);

		// Most languages include a constant for PI; C's should be M_PI
		// If you forget or it's not defined, just use however many digits
		// you can remember; at least 5 should be accurate enough.
		// PI / 2 is straight up (90 degrees)
		currentAngle = M_PI / 2; // 90 degrees
		// Calculate distance in radians between vertices
		angleSpacing = (2 * M_PI) / (points * 2); // 360 / (points * 2)

		// Calculate each vertex
		for(int i = 0; i < (points * 2); i++){
			// Use the appropriate radius - even i values represent points
			// on the outer circle
			if(i % 2 == 0){
				x = (outerRadius * cos(currentAngle)) + centerX;
				y = (outerRadius * sin(currentAngle)) + centerY;
			}
			else{
				x = (innerRadius * cos(currentAngle)) + centerX;
				y = (innerRadius * sin(currentAngle)) + centerY;
			}

			// Adjust for -0 issues
			// Values such as -0.001 will be rounded to (and print as)
			// -0.00, which isn't technically a number, and will
			// invalidate your response.
			if(x < 0 && x > -0.005){
				x = 0.00;
			}
			if(y < 0 && y > -0.005){
				y = 0.00;
			}

			// Print coordinates
			if(i != 0){
				printf(" ");
			}
			printf("%.2f,%.2f", x, y);

			// Move to the next angle; don't worry about exceeding 2*PI
			currentAngle = currentAngle + angleSpacing;
		}
		printf("\n");

		testCases = testCases - 1;
	}

	// close the file
	fclose(inputFile);
}

File: past-competitions/2018/src/brett-python/Prob16.py
import math
# Open the input file
with open("Prob16.in.txt", "rt") as inputFile:
    # Read the number of test cases (trim out the newline)
    cases = int(inputFile.readline().replace("\n", ""))
    # Set "constants"
    radians = math.pi * 2
    startAngle = math.pi / 2
    # For each test case
    for caseNum in range(cases):
        # Read the next line
        data = inputFile.readline().replace("\n", "")
        # Split the line
        pieces = data.split(" ")
        centerX = int(pieces[0])
        centerY = int(pieces[1])
        points = int(pieces[2])
        outerRadius = int(pieces[3])
        innerRadius = int(pieces[4])
        # Set starting angle and calculate distance between vertices
        currentAngle = startAngle
        angleDelta = radians / (points * 2)
        # Calculate vertices
        for i in range(points * 2):
            # Determine which radius to use; even i == points/outer
            radius = outerRadius
            if i % 2 == 1:
                radius = innerRadius
            # Calculate X and Y using given formulae
            vertexX = (radius * math.cos(currentAngle)) + centerX
            vertexY = (radius * math.sin(currentAngle)) + centerY
            # Correct for -0 errors; -0.001 will round to -0.00,
            # which is not a valid number
            if vertexX < 0 and vertexX > -0.005:
                vertexX = 0
            if vertexY < 0 and vertexY > -0.005:
                vertexY = 0
            # Print space as needed
            if i > 0:
                print(" ", end="")
            # Print coordinates (specify end to suppress newline,
            # and use format to force two decimal places)
            print("{:.2f},{:.2f}".format(vertexX, vertexY), end="")
            # Move to next angle (don't worry about exceeding 2PI)
            currentAngle = currentAngle + angleDelta
        # print newline when done
        print("")
        
